### ISSUE TITLE
"defaults" parameter in UrlRule doesn't work as described

### ISSUE THREAD

**Issue by OndrejVasicek** [yiisoft/yii2#17257]
I use the “defaults” parameter in UrlRule and it’s a great feature, but now I realized it doesn’t work in every promoted case.

This issue might be related with this #10970 but I’ll put it more straight – the functionality described in the guide (https://www.yiiframework.com/doc/guide/2.0/en/runtime-routing#default-parameter-values) doesn’t work in 50% of cases.

I started plain project with Yii 2.0.17. I put the only rule in rules in config as described in the guide

```php
'rules' => [
    // ...other rules...
    [
        'pattern' => 'posts/<page:\d+>/<tag>',
        'route' => 'post/index',
        'defaults' => ['page' => 1, 'tag' => ''],
    ],
]
```

In test controller I printed some test data. Exactly line in the guide.

```php
echo Url::to(['post/index',]);
echo '<br>';
echo Url::to(['post/index', 'page' => 2,]);
echo '<br>';
echo Url::to(['post/index', 'page' => 2, 'tag' => 'news',]);
echo '<br>';
echo Url::to(['post/index', 'tag' => 'news',]);
```

This is the result: (I normally don’t show index.php, but for the testing let’s keep it the same as the guide examples.)

```
/index.php/post/index
/index.php/posts/2
/index.php/posts/2/news
/index.php/post/index?tag=news
```

Only the second the third example worked. The first one and the last one didn’t. You can see it in the part “post” vs “posts”.
In the first example the result should be - `/index.php/posts`
In the last example the result should be - `/index.php/posts/news`
as described in the guide.

What’s wrong here? Am I missing something?
Thank you.

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.17
| PHP version      | 7.1.4
| Operating system | Win10

**Comment by OndrejVasicek** [comment]
Well, isn’t that a two-sided process? Why would one rule be able to parse an url without creating it? I can’t imagine a situation, where I can parse an url which I’m actually not able to create. It doesn’t make a sense to me. Why would I need to parse something which doesn’t exist?
Mainly – why it works in the two cases? And I have to say it’s very useful.

**Comment by rob006** [comment]
> Well, isn’t that a two-sided process?

It is, and now it is symmetric - parsing URL will give you route and GET params. If you try to create URL using the same route and params, you will get initial URL. 
Your case is different - you expect that this route with different params will still create the same URL.

**Comment by OndrejVasicek** [comment]
It doesn’t make sense to parse an url which will never exist. Why would I want to parse `/index.php/posts`, `/index.php/posts/2`, `/index.php/posts/2/news` and `/index.php/posts/news` urls by one UrlRule if they will never be created (by this rule)?

Let’s say I want to create the `/index.php/posts/news`. How can I do it? Well I have to create a new rule. And if I create a special rule for this URL, it handles even the parsing, so I don’t need the first rule with `defaults`. 

I must missing something, because there is no case this is useful as it works now.
But on the other hand – it would be very useful and desirable if the `defaults` worked as I expected, wouldn’t it?

**Comment by rob006** [comment]
> It doesn’t make sense to parse an url which will never exist. Why would I want to parse `/index.php/posts`, `/index.php/posts/2`, `/index.php/posts/2/news` and `/index.php/posts/news` urls by one UrlRule if they will never be created (by this rule)?

It will be created, you just need to provide correct params on URL creation.

> Let’s say I want to create the /index.php/posts/news. How can I do it?

Param in pattern will be skipped if you use default value. Since default `page` is `1`, you need to use `1` on URL creation:

```php
Url::to(['post/index', 'tag' => 'news', 'page' => 1])
```

And as I said, the simplest way to figure out which params you should use to create such URL, is to parse this URL and check what GET params you get.

**Comment by samdark** [comment]
Closed as duplicate.

**Comment by OndrejVasicek** [comment]
Thanks Rob for the explanation. Now I get it. Maybe it’s my level of English, but the guide doesn’t describe the way it works clearly. For me it suggest the functionality I described and which makes more sense to me (it’s more useful). Do you think it could be described more like you did?
I still have some (hope) valid points, which should be considered in order to keep the Yii great as possible.


1.
Despite the current functionality, which is great for situations like “hide pages 1 in url”, the proposed way of working would be also highly appreciated – it can dramatically reduce the number of URL rules, which is desirable goal according to the guide – why not to make the UrlRule work in the both ways as I suggested? I think it can work together, can’t see any drawback.

2.
The example with `page => 1` makes perfect sense and it’s useful. Can see it now. But the second case, where `tag => ''` seems to be nonsense to me. Why would I want to hide `tag` parameter from the URL using `defaults` in case there is no `tag` parameter passed? If there is no `tag` parameter passed, simply different UrlRule is applied. It actually creates this weird situation where these two URL creations
```php
Url::to(['post/index', 'tag' => '',]);
Url::to(['post/index']);
```
need two separate url rules to create the same url, only because the rule with `defaults` requires the `tag` parameter set, even if it’s empty.
The long story short – I don’t think it’s good to have more rules creating the same url. Because in this example one url is created by one rule but parsed by other.

**Comment by rob006** [comment]
> need two separate url rules to create the same url

Nope - both examples should match this rule:

```
    [
        'pattern' => 'posts/<tag>',
        'route' => 'post/index',
        'defaults' => ['tag' => ''],
    ],
```

But the problem is that you're trying to adjust your rules to match your `Url::to()` calls while you should do exactly the opposite - your `Url::to()` should be adjusted to rules definitions and params they're using.
I also encourage you to read #10970 - basically everything from this issue was already discussed there.

**Comment by OndrejVasicek** [comment]
Ok Rob, I went through the whole mentioned topic and even if I think it’s a bit clumsy and the docs would be better rewritten, I get it now. I adjusted the params in my creations to fit the rules.
But there is still something fishy. Consider this rule


```php
[
    'pattern' => '<controller:[\w\-]+>/<action:[\w\-]+>/<page:\d+>',
    'route' => '<controller>/<action>',
    'defaults' => ['action' => 'index', 'page' => 1,],
],
```

The purpose is quite clear – a general rule which “hides” two most common parameters in url for many situations – `action` in case it’s `index` and `page` in case it’s `1`. So instead of `product/index/1` I get `product` and instead of `product/index/2` I get `product/2`. The creation proces works like a charm. For testing purposes, I have only this one rule in rules.

When I open `product`, it works. The problem is, when I try to open the url `product/2`, the rule isn’t able to handle it and throws 404. If I remove `action` or `page` parameter from the defaults, it can suddenly handle it. So it’s not a problem with one or other, but the combination.
The funniest thing is that the debugger shows in ‘router’ section that the rule WAS USED. But somehow it doesn’t work. Like the `action` parameter wasn’t set to `index`.

**Comment by rob006** [comment]
@OndrejVasicek Does it work if you use:

```php
    'pattern' => '<controller:[\w\-]+>/<action:[a-z]+>/<page:\d+>',
```

?

**Comment by OndrejVasicek** [comment]
Huh? Yes, it does. But how? What's wrong with my regex? 
And how is it possible the "a-z" suits even for the "-" sign? Are you a magician?

**Comment by rob006** [comment]
`[a-z]` should not match actions with `-` - you need to use `[a-z\-]` instead. 

> What's wrong with my regex?

`\w` accepts also numbers, so `2` from `product/2` URL is interpreted as action name (since your regexp allows for this).

**Comment by OndrejVasicek** [comment]
Ou, my mistake, I mismatched a controller and an action. It works as you said. Thank you a lot Rob for your time.
